docs/algorithms-knn.md

```diff
@@ -0,0 +1,259 @@
+
+* Step 1. For each $k  \in [k_{min},k_{max}] $ ( $k_{min}$ has default value 1,  $k_{max}$has default value 100) in parallel select relevant features using FS+BRACE.
+* Step 2. Output the combination of features and  with the smallest LOOCV error.
+
+#### 6. Prediction
+If a target variable is available, we can compute the predicted target value for a query case q based on its k nearest neighbors.
+
+* Categorical target
+
+  Suppose that the target variable of the training cases has $J$ categories $1,2\ldots,J$, . Themajority voting will be used to predict the target value of a query case  based on the target values of its nearest neighbors, i.e.
+  $$
+  \widehat{y}_{q} =arg\max_{i}\left\{\sum_{i \in Nearest(q)}I(y_i=j), j=1,2,\ldots,J\right\}
+  $$
+
+
+where ) is a set containing the indices of k nearest neighbor of , and  is an indicator function which takes value 1 if , and 0 if .
+
+* Continuous target
+  For a query case , mean or  median of the target values among its  nearest neighbors will be used to predict the target value of q.
+
+#### 7. Feature importance
+Suppose there are m  features$X_{(1)},X_{(2)},\ldots,X_{(m)}$   in the final model. Then the feature importance will be computed using the leave-one-out method described as follows:
+
+* Step 1: $i = 1$
+
+* Step 2: Delete the feature $X_{(i)}$ from the model, and build kNN model based on other m-1  features, compute the classification error rate if target is categorical or sum-of-squares error if the target is continuous, denote it as $e_{(i)}$.
+
+* Step 3. If $i<m$, let $i=i+1$ , and go to step 2. Otherwise, go to step 4
+
+
+  $$
+
+  FI_{(i)} = {e_{(i)} + {1 \over m} \over \sum_{j=1}^m e_{(j) + 1} }, i=1,2,\ldots,m
+
+  $$
+
+
+## Returns
+The script outputs the indices of the k nearest neighbours of the query points. Additionally, if requested, the following returns will be provided: predicted values, selected k values, indices of the selected features, and feature importance.
+
+##Reference
+* Arya and Mount (1993). Algorithms for fast vector quantization.  In Proceedings of the Data Compression Conference, pages 381--390.
+* Andoni, A. and Indyk,P. (2008). Near-optimal hashing algorithm for approximate nearest neighbor in high dimensions. Communications of the ACM, 117-122.
+* Hamerly G. and Speegle G. (2010) Efficient Model Selection for Large-Scale Nearest-Neighbor Data Mining. BNCOD, 6121 of Lecture Notes in Computer Science, 37-54.
+* Li Y. and Lu B.-L.. (2009) Feature selection based on loss-margin of nearest neighbor classification. Pattern Recognition 42(9), 1914-1921.
+* Moore A. and Lee M. S. (1994) Efficient Algorithms for Minimizing Cross Validation Error. In International Conference on Machine Learning. 190-198.
+* Navot A., Shpigelman L., Tishby N. and Vaadia E. (2005) Nearest Neighbor Based Feature Selection for Regression and its Application to Neural Activity. In NIPS. 996-1002.
```

Comment:
This reference descibes a gradient-based method for selecting relevant features. This method is not used in the current implementation of kNN and this reference can be removed from the documentation.